### PR TITLE
Stop generating bad events

### PR DESCRIPTION
**Pull request by oguzhanunlu**
This PR fixes event generators to produce valid events that pass schema validation and adds comprehensive validation tests.

  ### Fix generators to produce valid events

  - Remove random vendor generation, use valid Snowplow endpoints only (tp1/tp2)
  - Fix IPv6 to use unsigned 16-bit range (0-65535) instead of signed Short
  - Fix geo_latitude (-90 to 90), geo_longitude (-180 to 180), geo_timezone
  - Fix Url port range typo (65335 -> 65535)
  - Make se_category and se_action required for structured events
  - Use realistic values for video dimensions, color depths, playback rates
  - Use valid ISO 4217 currency codes, AWS ID formats, email/phone formats
  - Fix OptimizelyState negative variationIndex, LinkClick elementClasses
  - Fix UaParserContext required fields, remove invalid YauaaContext IE fields

  ### Add validation tests

  - **SchemaValidationSpec** (62 tests): validates all contexts and unstructs against Iglu Central schemas
  - **AtomicFieldsValidationSpec** (100 tests): validates atomic field lengths against Enrich limits
    - Length constraints for all 84 string fields
    - Edge cases for event type-specific fields (struct, unstruct, transaction, etc.)
    - Format validation for currency codes, URL schemes, UUID
    - Tight constraint validation (geo_country, geo_region, base_currency)

ref: https://snplow.atlassian.net/browse/PDP-2396

### CODE REVIEW

File: core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/collector/Api.scala [hunk 1]
```diff
@@ -25,29 +25,16 @@ final case class Api(vendor: String, version: String) {
 }
 
 object Api {
-  private val GenI   = Gen.const(Api("i", ""))
-  private val GenIce = Gen.const(Api("ice", ".png"))
+  private val GenTp1 = Gen.const(Api("com.snowplowanalytics.snowplow", "tp1"))
```
Comment (benjben) [inline core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/collector/Api.scala]: Do you think that `tp1` brings value? `event-generator` is now used mostly for load testing. If we need to make sure that `/tp1` path works this needs to get covered by unit tests or Snowman.

File: core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/primitives/IpAddress.scala [hunk 2]
```diff
@@ -36,20 +36,22 @@ object IpAddress {
...
-        h <- Arbitrary.arbitrary[Short]
+        a <- hexSegment
+        b <- hexSegment
+        c <- hexSegment
+        d <- hexSegment
+        e <- hexSegment
+        f <- hexSegment
+        g <- hexSegment
+        h <- hexSegment
       } yield IpAddressV6(f"$a%x:$b%x:$c%x:$d%x:$e%x:$f%x:$g%x:$h%x")
```
Comment (benjben) [inline core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/primitives/IpAddress.scala:51]: This is fine but an IPv6 can also contain letters.

File: core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/protocol/contexts/EcommerceProduct.scala [hunk 3]
```diff
@@ -38,6 +38,6 @@ object EcommerceProduct extends SelfDescribingJsonGen {
       "brand"            -> strGen(1, 200).optionalOrNull,
       "inventory_status" -> strGen(1, 200).optionalOrNull,
       "position"         -> Gen.chooseNum(0, 9999999).optionalOrNull,
-      "currency"         -> strGen(3, 3).required
+      "currency"         -> Gen.oneOf("USD", "EUR", "GBP", "JPY", "AUD", "CAD", "CHF", "CNY", "SEK", "NZD").required
```
Comment (benjben) [inline core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/protocol/contexts/EcommerceProduct.scala:41]: I like it that `event-generator` generates events that are more real!

File: build.sbt [hunk 4]
```diff
@@ -43,7 +43,10 @@ lazy val core = project
       // Scala (test only)
       Dependencies.Libraries.specs2Scalacheck,
       Dependencies.Libraries.specs2,
-      Dependencies.Libraries.specs2Cats
+      Dependencies.Libraries.specs2Cats,
+      Dependencies.Libraries.igluClient,
+      Dependencies.Libraries.catsEffect,
```
Comment (benjben) [inline build.sbt:48]: This should be brought by `fs2` already?

File: core/src/test/scala/com/snowplowanalytics/snowplow/eventgen/SchemaValidationSpec.scala [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+import java.net.URI
+
+/** Validates that all context and unstruct event generators produce JSON that passes schema validation against Iglu
+  * Central.
+  */
+class SchemaValidationSpec extends CatsResource[IO, IgluCirceClient[IO]] with SpecificationLike {
+  import SchemaValidationSpec._
+
+  skipAllIf(SchemaValidationSpec.javaVersion < 11)
```
Comment (benjben) [inline core/src/test/scala/com/snowplowanalytics/snowplow/eventgen/SchemaValidationSpec.scala]: It's not common to see this. Who/what could be using Java < 11?

**Review by benjben — APPROVED** [review]
Thanks @oguzhanunlu!

I really like how the events are closer to real events!

File: core/src/test/scala/com/snowplowanalytics/snowplow/eventgen/SchemaValidationSpec.scala (hunk 5 above)
Comment (oguzhanunlu) [reply core/src/test/scala/com/snowplowanalytics/snowplow/eventgen/SchemaValidationSpec.scala]: > Who/what could be using Java < 11?

this project :) https://github.com/snowplow-incubator/snowplow-event-generator/blob/70df176df3ca25197f3d56f9293e7d0ae666c439/.github/workflows/publish.yml#L17

We can drop Java 8 support from CI and get rid of this version check altogether. I don't foresee any problem with it, do you?

File: core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/primitives/IpAddress.scala (hunk 2 above)
Comment (oguzhanunlu) [reply core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/primitives/IpAddress.scala:51]: yeah, I can improve it

File: core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/collector/Api.scala (hunk 1 above)
Comment (oguzhanunlu) [reply core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/collector/Api.scala]: that's fair, I can get rid of `/tp1`

File: core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/primitives/IpAddress.scala (hunk 2 above)
Comment (oguzhanunlu) [reply core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/primitives/IpAddress.scala:51]: I just noticed that `%x` already gives us lowercase letters.

```scala
scala> f"${65535}%x"
res0: String = ffff
```

File: build.sbt (hunk 4 above)
Comment (oguzhanunlu) [reply build.sbt:48]: `core` project doesn't have fs2, `sinks` do

File: core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/primitives/IpAddress.scala (hunk 2 above)
Comment (benjben) [reply core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/primitives/IpAddress.scala:51]: I had missed that sorry, perfect then!

File: core/src/test/scala/com/snowplowanalytics/snowplow/eventgen/SchemaValidationSpec.scala (hunk 5 above)
Comment (benjben) [reply core/src/test/scala/com/snowplowanalytics/snowplow/eventgen/SchemaValidationSpec.scala]: We should drop it indeed!

**Review by benjben — APPROVED** [review]
